### DISCUSSION TITLE
emsdk built on emsdk

### DISCUSSION THREAD

**Discussion by ericoporto** [emscripten-core/emscripten#16329] [Ideas]
This is kinda ridiculous question and I don't know if it has been done. 

I was looking the Apple arm64 issue the other day and I had a thought... Would be possible to build all dependencies of emsdk in emsdk so they could be run by some WASM/JS engine so one could access the Emscripten tools from any platform with a working WASM/JS runner?

**Comment by sbc100** [comment]
This has come up before, and its that same questions as "can I run tool X as a wasm binary".. where X could be clang or xcode or any developer tool (vim, emacs, etc).

@kripken is a big fan if pioneering this approach, but its seems little "everything looks like a nail" to me.  If we are going to start running command line tools based on wasm... I would expect other, simpler, command line projects to lead they way rather than a complex, multi-binary, multi-process, multi-language, filesystem-heavy project like emscripten.

**Reply by kripken** [reply]
There are simpler commandline projects, but they might not be familiar enough with wasm to think it is even possible...

The reason why I think this is not the "nail" situation is that this has actual benefits. A single wasm build that runs at 90% of speed would be good enough, and much simpler than building for multiple OSes, both for us as the builders and also for users that happen to use a platform we don't build for.

**Reply by sbc100** [reply]
Yes, the benefits of a single build are well known.. but those benefits apply equally well to literally all C/C++ codebases that want to run in more than one place (effectively all of them).   I guess it makes some sense that emscripten might want to pioneer this, but I still think it would make sense to choose something simpler first.

**Comment by kripken** [comment]
See https://github.com/jprendes/emception for the latest effort there - that even runs in a browser, which is harder than in Node.js. So definitely doable.

I do think this is very useful to do for Node.js and wasm VMs, for portability. However, there are issues to investigate, like https://github.com/emscripten-core/emscripten/issues/15727, and also, to be fast the VM would need pthreads + EH support, which not all Node VMs do atm, so it might be too early.
